mlir/lib/Bindings/Python/IRCore.cpp

```diff
@@ -1249,6 +1250,21 @@ void PyOperationBase::writeBytecode(const py::object &fileObject,
                               .str());
 }
 
+void PyOperationBase::walk(
+    std::function<MlirWalkResult(MlirOperation)> callback,
+    MlirWalkOrder walkOrder) {
+  PyOperation &operation = getOperation();
+  operation.checkValid();
+  MlirOperationWalkCallback walkCallback = [](MlirOperation op,
+                                              void *userData) {
+    auto *fn =
+        static_cast<std::function<MlirWalkResult(MlirOperation)> *>(userData);
+    return (*fn)(op);
```

Comment:
thought: can you actually check the result here to see if it's `py::none()` (i.e., the callback doesn't have a `return` statement) and then default to `MlirWalkResult::MlirWalkResultAdvance`. I think that's reasonable semantics? @ftynse

Reply:
I'm not sure it works since the function is type casted to  `std::function<MlirWalkResult(MlirOperation)> callback`. 
To allow void functions I think we could change the callback type to `std::variant<std::function<MlirWalkResult(MlirOperation)>, std::function<void(MlirOperation)>` or maybe `py::function` (and dynamically type casts the returned type).

Reply:
variant is overkill for sure but `py::function` might work? anyway it's just a thought rather than a hard request so nbd if you don't want to bother.

Reply:
My two cents is if we can do this in a way that doesn't add much overhead it would be nice. But I'm also happy to see this go in with the user callback required to explicitly return a WalkResult... "explicit is better than implicit".

Reply:
I guess I'm too traumatized by the C++ version of omitting the return from a function and getting a random segfault, so I'm on the side of always having an explicit return. But not objecting strongly.

Reply:
> I guess I'm too traumatized by the C++ version of omitting the return from a function and getting a random segfault

I am similarly traumatized but in Python-land I think it's fairly well-known that a function with no return statement actually returns `None`.